qubes/tests/vm/mix/net.py

```diff
@@ -159,6 +159,76 @@ def test_145_netvm_change(self):
                 mock_detach.reset_mock()
                 mock_attach.reset_mock()
 
+    def test_146_netvm_defer(self):
+        vm = self.get_vm()
+        self.setup_netvms(vm)
+        with (
+            patch("qubes.vm.qubesvm.QubesVM.is_running", lambda x: True),
+            patch("qubes.vm.qubesvm.QubesVM.is_paused", lambda x: True),
+            patch("qubes.vm.mix.net.NetVMMixin.attach_network") as mock_attach,
+            patch("qubes.vm.mix.net.NetVMMixin.detach_network") as mock_detach,
+            patch("qubes.vm.qubesvm.QubesVM.create_qdb_entries"),
+        ):
+
+            with self.subTest("changing netvm"):
+                original_netvm = vm.netvm.name
+                vm.netvm = self.netvm2
+                mock_detach.assert_not_called()
+                mock_attach.assert_not_called()
+                self.assertEqual(
+                    vm.features.get("deferred-netvm-original", None),
+                    original_netvm,
+                )
+                with patch(
+                    "qubes.vm.qubesvm.QubesVM.is_paused", lambda x: False
+                ):
+                    vm.apply_deferred_netvm()
+                    self.assertEqual(
+                        vm.features.get("deferred-netvm-original", None), None
+                    )
+                    mock_detach.assert_called()
+                    mock_attach.assert_called()
+                    mock_detach.reset_mock()
+                    mock_attach.reset_mock()
+
+            with self.subTest("setting netvm to none"):
+                original_netvm = vm.netvm.name
+                vm.netvm = None
+                mock_detach.assert_not_called()
+                mock_attach.assert_not_called()
+                self.assertEqual(
+                    vm.features.get("deferred-netvm-original", None),
+                    original_netvm,
+                )
+                with patch(
+                    "qubes.vm.qubesvm.QubesVM.is_paused", lambda x: False
+                ):
+                    vm.apply_deferred_netvm()
+                    self.assertEqual(
+                        vm.features.get("deferred-netvm-original", None), None
+                    )
+                    mock_detach.assert_called()
+                    mock_attach.assert_not_called()
+
+            with self.subTest("resetting netvm to default"):
+                original_netvm = vm.netvm.name if vm.netvm else ""
+                del vm.netvm
+                mock_detach.assert_called()
```

Comment:
I don't think it should be called on paused vm. It looks like you are simply missing reset_mock() calls after previous test.